### PR TITLE
Add gtests for MPPTunnel

### LINKED ISSUE #4481: Add GTest for MPPTunnel
## Enhancement
Add gtest to ensure the code quanlity and make it convenient to do code refact later.

### PR DESCRIPTION
**Pull request by yibin87**
Signed-off-by: yibin <huyibin@pingcap.com>

### What problem does this PR solve?

Issue Number: close #4481 

Problem Summary:

### Check List

Tests 

- [ x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by yibin87** [comment]
Previous PR seems breaks when TiFlash make public, so create this new one. And add is_local and is_async tests.

**Review by bestwoody — APPROVED** [review]

File: dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp [hunk 1]
```diff
@@ -0,0 +1,645 @@
...
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockWriter>();
+    mpp_tunnel_ptr->connect(writer_ptr.get());
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), true);
+    std::unique_ptr<mpp::MPPDataPacket> data_packet_ptr = std::make_unique<mpp::MPPDataPacket>();
+    data_packet_ptr->set_data("First");
+    mpp_tunnel_ptr->write(*data_packet_ptr);
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000)); // wait for sendJob thread to write data
+    mpp_tunnel_ptr->close("Canceled");
```
Comment (windtalker) [inline dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp:313]: Why not call `close` with empty string? This is the normal case and no error message will be written to the writer.

File: dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockWriter>();
+    mpp_tunnel_ptr->connect(writer_ptr.get());
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), true);
+    std::unique_ptr<mpp::MPPDataPacket> data_packet_ptr = std::make_unique<mpp::MPPDataPacket>();
+    data_packet_ptr->set_data("First");
+    mpp_tunnel_ptr->write(*data_packet_ptr, true);
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000)); // wait for sendJob thread to write data
+    mpp_tunnel_ptr->waitForFinish();
```
Comment (windtalker) [inline dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp:331]: If `waitForFinish` is called, then there is no need to sleep for the sendJob thread?

File: dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp (hunk 1 above)
Comment (windtalker) [reply dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp:313]: And I think `close` will blocked until the sendJob thread finishes, so no need to sleep?

File: dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockWriter>();
+    mpp_tunnel_ptr->connect(writer_ptr.get());
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), true);
+    auto data_packet_ptr = std::make_unique<mpp::MPPDataPacket>();
+    data_packet_ptr->set_data("First");
+    mpp_tunnel_ptr->write(*data_packet_ptr);
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000)); // wait for sendJob thread to write data
+    mpp_tunnel_ptr->writeDone();
```
Comment (windtalker) [inline dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp:349]: Ditto, no need to sleep

File: dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+try
+{
+    auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
+    std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockWriter>();
+    mpp_tunnel_ptr->connect(writer_ptr.get());
+    GTEST_ASSERT_EQ(mpp_tunnel_ptr->getConnectFlag(), true);
+    auto data_packet_ptr = std::make_unique<mpp::MPPDataPacket>();
+    data_packet_ptr->set_data("First");
+    mpp_tunnel_ptr->write(*data_packet_ptr);
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000)); // wait for sendJob thread to write data
```
Comment (windtalker) [inline dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp:366]: Ditto

File: dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+
+public:
+    std::vector<String> write_packet_vec;
+};
+
+class MockFailedWriter : public PacketWriter
+{
+    bool write(const mpp::MPPDataPacket & packet) override
+    {
+        write_packet_vec.push_back(packet.data());
```
[the remark this answers is not in the document]
Comment (yibin87) [reply dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp:80]: To avoid unused paremeter, anyway, I'll remove so to see if format complain anything.

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**